File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.go
func largestRectangleArea(heights []int) int {
    st := make([]int, 0)
    st = append(st, -1);
    n := len(heights)
    ans := 0

    for i, h := range heights {
        for len(st) > 1 && heights[st[len(st)-1]] > h {
            t := st[len(st)-1]
            st = st[:len(st)-1]
            w := i - st[len(st)-1] - 1
            area := w * heights[t]
            ans = max(ans, area)
        }
        st = append(st, i)
    }

    for len(st) > 1 {
        t := st[len(st)-1]
        st = st[:len(st)-1]
        w := n - st[len(st)-1] - 1
        area := w * heights[t]
        ans = max(ans, area)
    }

    return ans
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        st.push(-1);

        int ans = 0;
        for (int i = 0; i < heights.size(); i++) {
            while(st.size() > 1 && heights[st.top()] > heights[i]) {
                int t = st.top();
                st.pop();
                int weight = i - st.top() - 1;
                int curr = heights[t] * weight;
                ans = max(ans, curr);
            }

            st.push(i);
        }

        while(st.size() > 1) {
            int t = st.top();
            st.pop();
            int weight = heights.size() - st.top() - 1;
            int curr = heights[t] * weight;
            ans = max(ans, curr);
        }

        return ans;
    }
};